File: checksum.py
import re, sys, hashlib
from rich.console import Console

def identify_hash(s: str) -> str:
    """
    Identifies the hash type of a given string (s) based on its length
    and character set

    Args:
    s (str): a hash value of type string

    Returns:
    str: the hash type of s
    """
    if len(s) == 32 and re.match(r'^[a-fA-F0-9]{32}$', s):
        return "md5"
    elif len(s) == 40 and re.match(r'^[a-fA-F0-9]{40}$', s):
        return "sha1"
    elif len(s) == 56 and re.match(r'^[a-fA-F0-9]{56}$', s):
        return "sha224"
    elif len(s) == 64 and re.match(r'^[a-fA-F0-9]{64}$', s):
        return "sha256"
    elif len(s) == 96 and re.match(r'^[a-fA-F0-9]{96}$', s):
        return "sha384"
    elif len(s) == 128 and re.match(r'^[a-fA-F0-9]{128}$', s):
        return "sha512"
    else:
        return ""

def calculate_hash(file_path: str, hash_type: str) -> str:
    """
    Calculates the hash for a file.

    Args:
    file_path (str): Path to the file.
    hash_type (str): The type of hash to calculate for the file.

    Returns:
    str: hash of the file.
    """

    if hash_type == "md5":
        file_hash = hashlib.md5()
    elif hash_type == "sha1":
        file_hash = hashlib.sha1()
    elif hash_type == "sha224":
        file_hash = hashlib.sha224()
    elif hash_type == "sha256":
        file_hash = hashlib.sha256()
    elif hash_type == "sha384":
        file_hash = hashlib.sha384()
    elif hash_type == "sha512":
        file_hash = hashlib.sha512()
        
    with open(file_path, "rb") as f:
        chunk = 0
        while chunk := f.read(4096):
            file_hash.update(chunk)
    return file_hash.hexdigest()

def main():
    console = Console()
    if (len(sys.argv) -1) > 2:
        console.print("[bold red]Too many args\nUsage: python3 checksum.py /path/to/file hash_value[/bold red]")
    elif (len(sys.argv) -1) < 2:
        console.print("[bold red]Too few args\nUsage: python3 checksum.py /path/to/file hash_value[/bold red]")
    else:
        file_path = sys.argv[1]
        hash_value = sys.argv[2]
        try:
            hash_type = identify_hash(hash_value)
            console.print(f"[bold green]Using {hash_type.upper()} Algorithm...[/bold green]")
            checksum = calculate_hash(file_path, hash_type)
            if checksum == hash_value:
                out = f"[bold green]File hash is authentic\nProvided hash: {hash_value}\nCalculated hash: {checksum}[/bold green]"
            else:
                out = f"[bold red]File hash IS NOT authentic\nProvided hash: {hash_value}\nCalculated hash: {checksum}[/bold red]"
            console.print(out)
        except FileNotFoundError:
            print("File not found!")
        except IsADirectoryError:
            print("Given path is a directory, please provide a file path.")
        except Exception as e:
            print(f"An error occurred: {e}")
    
if __name__ == "__main__":
    main()

File: checksum.go
package main

import (
	"fmt"
	"os"
	"strings"
	"hash"
	"encoding/hex"
	"crypto/sha256"
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha512"
)

var Reset = "\033[0m" 
var Red = "\033[31m" 
var Green = "\033[32m"
var Yellow = "\033[33m"

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func IdentifyHashType(hash string) string {
	if len(hash) == 32 {
		return "md5"
	} else if len(hash) == 40 {
		return "sha1"
	} else if len(hash) == 56 {
		return "sha224"
	} else if len(hash) == 64 {
		return "sha256"
	} else if len(hash) == 96 {
		return "sha384"
	} else if len(hash) == 128 {
		return "sha512"
	} else {
		return ""
	}
}

func CalculateFileHash(filePath string, hashType string) (string, error) {
	dat, err := os.ReadFile(filePath)
	check(err)

	var h hash.Hash

	switch hashType {
	case "md5":
		h = md5.New()
	case "sha1":
		h = sha1.New()
	case "sha224":
		h = sha256.New224()
	case "sha256":
		h = sha256.New()
	case "sha384":
		h = sha512.New384()
	case "sha512":
		h = sha512.New()
	default:
		return "", fmt.Errorf("unsupported hash type: %s", hashType)
	}

	h.Write(dat)
	hashBytes := h.Sum(nil)

	return hex.EncodeToString(hashBytes), nil
}

func main() {
	if len(os.Args) < 3 {
		fmt.Println(Red + "Usage: checksum <file_path> <hash>" + Reset)
		fmt.Println(Red + "Issues @ https://github.com/ibnaleem/checksum/issues" + Reset)
		os.Exit(1)
	} else {
		filePath := os.Args[1]
		hashValue := os.Args[2]
		hashType := IdentifyHashType(hashValue)
		
		if hashType == "" {
			fmt.Println(Red + "Unsupported hash type provided" + Reset)
			os.Exit(1)
		}
		
		fmt.Println(Yellow + "Using " + strings.ToUpper(hashType) + " algorithm to calculate hash for " + filePath + "..." + Reset)
		calculatedHash, err := CalculateFileHash(filePath, hashType)
		
		if err != nil {
			fmt.Printf("%s Error: %s %s", Red, err, Reset)
		} else {
			if calculatedHash == hashValue {
				fmt.Printf(Green + "Checksum matches for %s using %s algorithm!\n", filePath, strings.ToUpper(hashType))
				fmt.Println(Green + "Provided hash: " + hashValue)
				fmt.Println(Green + "Calculated hash: " + calculatedHash + Reset)
			} else {
				fmt.Printf(Red + "Checksum does not match for %s using %s algorithm!\n", filePath, hashType)
				fmt.Println(Red + "Provided hash: " + hashValue)
				fmt.Println(Red + "Calculated hash: " + calculatedHash)
			}
		}
	}
}
